### PR TITLE
[world-local] Reduce stream metadata I/O

### PR DESCRIPTION
**Pull request by pranaygp**
## Summary
- read only the one-byte EOF marker when `getInfo()`, paginated `getChunks()`, or negative-index resolution do not need stream payload bytes
- scan the stream chunks directory once per lookup while preserving legacy, untagged, and tagged-file precedence
- add focused coverage for the first-byte filesystem helper and a patch changeset for `@workflow/world-local`

## Performance
Local micro-benchmark using a completed stream containing 64 x 1 MiB chunks, calling `getInfo()` 20 times:

| Version | Time per call |
| --- | ---: |
| Before | 16.40 ms |
| After | 3.40-3.58 ms |

This is approximately a 4.7x improvement for the measured metadata-read path.

## Testing
- `pnpm --filter @workflow/world-local build`
- `pnpm --filter @workflow/world-local test` (`8` test files, `352` tests passed)
- `pnpm exec biome check packages/world-local/src/fs.ts packages/world-local/src/fs.test.ts packages/world-local/src/streamer.ts .changeset/quieter-stream-metadata.md` (passes; reports existing complexity warnings in previously complex functions)
- `git diff --check`

### CODE REVIEW

**Review by TooTallNate — APPROVED** [review]
## Approve — clean perf win with no behavioral change

Three optimizations stacked in one PR, each correct on its own:

1. **`readFirstByte` helper** replaces `readBuffer(file)` in three EOF-check sites. The chunk format is `[eofByte] [payload]` (`serializeChunk` at `streamer.ts:60`), so reading byte 0 is sufficient to determine EOF. For multi-MB chunks this is the difference between reading 1 byte and reading the full payload.

2. **`listChunkEntries` + `addChunkFilesByExtension`** collapses 2–3 parallel `listFilesByExtension` calls (each doing its own `fs.readdir`) into one readdir + in-memory filtering. Saves 1–2 readdir syscalls per `listChunkFilesForStream` call.

3. **Eliminates `deserializeChunk` from the negative-startIndex path** — previously read the full last chunk just to check the EOF flag, now reads one byte. The "extra disk read" comment was honest about being slow; now it's gone.

## Behavioral equivalence verified

I traced `listChunkFilesForStream` precedence carefully (legacy `.json` → untagged `.bin` → tagged `.tag.bin`) with an example directory containing `chunk-01ABC.bin`, `chunk-02ABC.mytag.bin`, `chunk-03ABC.json`, with `tag='mytag'`. Both old and new produce the same `extMap`:

```
{ 'chunk-03ABC': '.json', 'chunk-01ABC': '.bin', 'chunk-02ABC': '.mytag.bin' }
```

The subtle part is that the old code's `binFiles` includes tagged files (because `.mytag.bin` ends with `.bin`), then filters them out via `if (tag && f.endsWith(tagSfx)) continue`. The new code's `addChunkFilesByExtension` accepts an `include` predicate that does the same check (`file => !file.endsWith('.${tag}')`). Identical outcome.

## What I verified locally

- `pnpm install --frozen-lockfile` ✓
- `pnpm turbo run build --filter @workflow/world-local` ✓
- `pnpm --filter @workflow/world-local test` ✓ (365 pass)
- New test `readFirstByte` reads only the marker value and handles empty files ✓
- Existing negative-startIndex tests (`should support negative startIndex to read from the end`, `should clamp negative startIndex that exceeds chunk count to 0`) both pass, exercising the new `readFirstByte` path through `getInfo()` / negative-index resolution
- Existing pagination tests in `getChunks` exercise the cursor-vs-EOF path

## Edge cases looked OK

- **Empty file** (`Buffer.alloc(0)`): `readFirstByte` returns `undefined`. The streamer checks `=== 1`, so empty files are treated as non-EOF data chunks. The old code's `isEofChunk(buffer)` on an empty buffer returns `false` (since `buffer[0]` is `undefined`). Same behavior.
- **`Buffer.allocUnsafe(1)`** is correct here because the byte is immediately overwritten by `file.read()`, and the `bytesRead === 0` branch returns `undefined` before reading from the buffer.
- **`fs.open` errors** (file deleted between readdir and open, permissions, etc.) propagate as exceptions — same as `readBuffer` did. Pre-existing race condition, not introduced.

## CI

37 success, 0 failures, 73 in progress. The optimization is internal to world-local so doesn't affect anything outside the package; expecting clean CI.

## Two small notes (not blockers)

1. **The `readFirstByte` test could exercise EOF-byte-0 too**:
   ```js
   await fs.writeFile(eofFalsePath, Buffer.from([0, 2, 3]));
   expect(await readFirstByte(eofFalsePath)).toBe(0);
   ```
   The current test only checks the value-1 case + empty file. A `===  0` case would document that the streamer's `=== 1` check correctly distinguishes EOF from non-EOF. Minor.

2. **`addChunkFilesByExtension` signature has 5 parameters with two having defaults** — readable but a bit clunky. A small refactor would be to take an options object:
   ```ts
   addChunkFilesByExtension(extMap, entries, {
     sourceExtension: '.bin',
     filter: tag ? (file) => !file.endsWith(`.${tag}`) : undefined,
   });
   ```
   But this is purely stylistic — the current call sites read fine in context.

Approving.

File: packages/world-local/src/streamer.ts [hunk 1]
```diff
@@ -333,7 +353,7 @@ export function createStreamer(basedir: string, tag?: string): Streamer {
 
           // Before the cursor: only need to check EOF (1 byte), skip content
           if (dataIndex < startIndex) {
-            if (isEofChunk(await readBuffer(filePath))) {
+            if ((await readFirstByte(filePath)) === 1) {
```
Comment (karthikscale3) [inline packages/world-local/src/streamer.ts]: Now that callers use inline `(await readFirstByte(...)) === 1`, the `isEofChunk` helper (line 49) has zero remaining callers and isn't re-exported — it's dead code. Suggest removing it, or better, repurposing it into a shared EOF helper (see comment on line 415) so the named abstraction survives instead of being orphaned.

File: packages/world-local/src/streamer.ts [hunk 2]
```diff
@@ -386,14 +406,13 @@ export function createStreamer(basedir: string, tag?: string): Streamer {
...
-        // Only read the first byte of each file to check EOF — no full
-        // deserialization needed since we just need a count.
+        // Read only the EOF marker byte because metadata never needs payloads.
         let streamDone = false;
         let dataCount = 0;
         for (const file of chunkFiles) {
           const ext = fileExtMap.get(file) ?? '.bin';
           if (
-            isEofChunk(await readBuffer(path.join(chunksDir, `${file}${ext}`)))
+            (await readFirstByte(path.join(chunksDir, `${file}${ext}`))) === 1
```
Comment (karthikscale3) [inline packages/world-local/src/streamer.ts]: The literal `1` as the EOF marker now appears in `serializeChunk`, `deserializeChunk`, and 5 new `=== 1` comparison sites. Consider a named constant + tiny helper to remove the magic number and give `isEofChunk` a reason to live:

```ts
export const EOF_MARKER = 1;
export const isEofByte = (b: number | undefined) => b === EOF_MARKER;
// usage: if (isEofByte(await readFirstByte(filePath))) { ... }
```

File: packages/world-local/src/streamer.ts [hunk 3]
```diff
@@ -68,25 +92,21 @@ async function listChunkFilesForStream(
...
-  const taggedBinFiles = taggedResults[0] ?? [];
-
+  const entries = await listChunkEntries(chunksDir);
   const extMap = new Map<string, string>();
-  for (const f of jsonFiles) extMap.set(f, '.json');
-  const tagSfx = tag ? `.${tag}` : '';
-  for (const f of binFiles) {
-    if (tag && f.endsWith(tagSfx)) continue;
-    extMap.set(f, '.bin');
+  addChunkFilesByExtension(extMap, entries, '.json');
```
Comment (karthikscale3) [inline packages/world-local/src/streamer.ts:103]: This single-scan refactor is the riskier change in the PR (precedence + tag-exclusion logic), but it only has transitive coverage via existing streamer tests. A small targeted test asserting the resulting `extMap` for a mixed directory (legacy `.bin`, `.json`, and `.<tag>.bin` entries) would lock in the behavior you carefully preserved here.

**Review by karthikscale3 — APPROVED** [review]
Approve with minor cleanups. Logic is correct and behavior-preserving: `firstByte === 1` is exactly equivalent to the old `isEofChunk`/`deserializeChunk().eof` given the `[1-byte eof][payload]` format, the single-`readdir` refactor preserves json→bin→tagged precedence and tag exclusion, and the two payload-needing paths were correctly left as full reads. Nice targeted ~4.7x win on the metadata path.

A few non-blocking nits left inline.

**Review by VaguelySerious — APPROVED** [review]

File: packages/world-local/src/streamer.ts (hunk 1 above)
Comment (pranaygp) [reply packages/world-local/src/streamer.ts]: Addressed in `21bfb77c2`: `isEofChunk()` now delegates to the shared EOF-byte predicate, and the marker-only disk probes use that same predicate.

File: packages/world-local/src/streamer.ts (hunk 2 above)
Comment (pranaygp) [reply packages/world-local/src/streamer.ts]: Addressed in `21bfb77c2`: introduced a named `EOF_MARKER` plus `isEofByte()` predicate and routed serialization, deserialization, and marker-only reads through them.

File: packages/world-local/src/streamer.ts (hunk 3 above)
Comment (pranaygp) [reply packages/world-local/src/streamer.ts:103]: Addressed in `21bfb77c2`: added a focused `getChunks()` test with colliding `.json`, `.bin`, and `.vitest-0.bin` entries, asserting tagged > untagged > legacy selection and EOF handling.
